### ISSUE TITLE
StableDiffusionXLInpaintPipeline.from_single_file for refiner returns back Img2ImgPipeline

### ISSUE THREAD

**Issue by darshats** [huggingface/diffusers#6001]
### Describe the bug

When using this call:
StableDiffusionXLInpaintPipeline.from_single_file(<path>/sd_xl_refiner_1.0.safetensors), we get an instance of Img2ImgPipeline rather than Inpainting pipeline.
This seems related to [4186](https://github.com/huggingface/diffusers/issues/4186), but in this case Img2ImgPipeline doesnt throw an error when mask_image parameter is passed, it just silently ignores it.

### Reproduction

from diffusers import StableDiffusionXLInpaintPipeline
from diffusers.utils import load_image

refiner = StableDiffusionXLInpaintPipeline.from_single_file(
    "./sd_xl_refiner_1.0.safetensors",
    torch_dtype=torch.float16,
    use_safetensors=True,
    variant="fp16",
)
refiner.to("cuda")

### System Info

Diffusers 0.23

**Comment by darshats** [comment]
The intent is to get a refiner for inpainting like described [here](https://github.com/huggingface/diffusers/blob/main/docs/source/en/using-diffusers/inpaint.md#inpaint-to-image-to-image)

However if I run the code below, `StableDiffusionXLInpaintPipeline.from_single_file` call returns an Img2Img pipeline. The `AutoPipelineForInpainting.from_pipe` returns an Inpainting pipeline. The single file call looks like a bug?


     pipe = StableDiffusionXLInpaintPipeline.from_single_file(
        "[https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors"](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors%22),
        torch_dtype=torch.float16,
        variant="fp16",
        use_safetensors=True
        )
    pipe.to("cuda")
 
    ### this returns in Img2Img pipeline that ignores the mask parameter
    refiner = StableDiffusionXLInpaintPipeline.from_single_file(
        "[https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/main/sd_xl_refiner_1.0_0.9vae.safetensors"](https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/main/sd_xl_refiner_1.0_0.9vae.safetensors%22),
        text_encoder_2=pipe.text_encoder_2,
        vae=pipe.vae,
        torch_dtype=torch.float16,
        variant="fp16",
        use_safetensors=True
    )
    refiner.to('cuda')
 
    ## this returns an inpainting pipeline
    refiner2 = AutoPipelineForInpainting.from_pipe(
        refiner,
        text_encoder_2=pipe.text_encoder_2,
        torch_dtype=torch.float16,
        variant="fp16",
        use_safetensors=True
        )
    refiner2.to('cuda')

**Comment by patrickvonplaten** [comment]
The issue is quite easy to reproduce:

```py
from diffusers import StableDiffusionXLInpaintPipeline
import torch

pipe = StableDiffusionXLInpaintPipeline.from_single_file(
    "https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors",
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True
)
refiner = StableDiffusionXLInpaintPipeline.from_single_file(
    "https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/main/sd_xl_refiner_1.0_0.9vae.safetensors",
    text_encoder_2=pipe.text_encoder_2,
    vae=pipe.vae,
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True
)


print("Refiner", refiner.__class__.__name__)   # prints incorrect class
```


@DN6 it would be nice if you can take a look here since we want to refactor the `from_single_file` function anyways

**Comment by darshats** [comment]
There is another similar issue - on diffusers 0.22.3
The call 
`pipe = StableDiffusionXLPipeline.from_single_file(
        "https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors", 
        torch_dtype=torch.float16, 
        variant="fp16", 
        use_safetensors=True
        )
 `       
returns back an object of type inpaint pipeline, but the unet config only has 4 channels.

If I use:
`
pipe = AutoPipelineForInpainting.from_pretrained(
        "diffusers/stable-diffusion-xl-1.0-inpainting-0.1", 
        torch_dtype=torch.float16, 
        variant="fp16"
        )
`
 it correctly returns back in inpaint pipeline instance, with unet having 9 channels.
 
Some docs have the first way of calling (in conjunction with refiner) and some docs have the second (autopipeline) way of calling. But the second one is correct. First one config looks suspect, and its not really inpainting due to unet config only 4 channels.

**Comment by DN6** [comment]
@darshats should be fixed with https://github.com/huggingface/diffusers/pull/6147
